rails_programming/hotwire/turbo.md

```diff
@@ -0,0 +1,494 @@
+...
+  def create
+    @post = Post.new(post_params)
+
+    respond_to do |format|
+      if @post.save
+        format.turbo_stream
+      else
+        format.html { render :new, status: :unprocessable_entity }
+      end
+    end
+  end
+~~~
+
+What's going on here? Well, the `respond_to do |format|` block is our way of telling the controller to do more than just `html` format requests. If our post saves, we would like the format we respond with to be a `create.turbo_stream.erb` format. If it does not save, we would like to respond with the `new.html.erb` file being rendered. Notice the difference in the formats is the file extension that is after the name and before the `.erb`.
+
+We include the `422 Unprocessable Entity` error code with our response upon an unsaved post so that Turbo Drive knows what to do with it and our form window won't disappear anymore!
+
+Now all we have to do is create our `create.turbo_stream.erb` file to respond with!
+
+#### Turbo Stream Template
+
+You create your Turbo Stream file inside of your `views` folder the same way as any other view. In this example, it would be located at `views/posts/create.turbo_stream.erb`. Our view file will look like this:
+
+~~~erb
+<%= turbo_stream.append "posts", @post %>
+~~~
+
+That line is all we need! What this does is create a Turbo Stream packet with the `append` action. The target of the action is `"posts"`. This is the `<div>` with the `id="posts"` that contains all of our posts. Turbo Stream will locate the `<div>` and append our brand new `@post` to the bottom! Our newly created `@post` will even use the `_post.html.erb` partial to be in the proper layout as all of our other posts. For that, you can thank the Rails naming convention. If you are breaking from Rails convention, you can specify a partial to use inside of your `turbo_stream` template.
+
+And just like that, we can create a post and watch it show up on the list!
+
+#### Additional Turbo Stream Tips
+
+There's no time to get more in-depth on how Turbo Stream works in this lesson, but here are some topics for you to begin your personal research
+if you would like to learn more:
+
+1.  You can chain multiple `turbo_stream` actions in one `turbo_stream.erb` file, like if you wanted to append to one region and
+    update another region at the same time.
+
+2.  You don't *have* to create a separate layout file for your Stream formats. For instance, we could instead change the line in our controller to:
+
+    ~~~ruby
+    format.turbo_stream { render turbo_stream: turbo_stream.append('posts', @post) }
+    ~~~
+
+    However, [you shouldn't do this for anything complex or chained](https://github.com/hotwired/turbo-rails/issues/77#issuecomment-757349251).
+
+3.  You may have noticed that when you submit a new Post, the text box doesn't clear out! That's why Hotwire comes with Stimulus. In the next lesson, we will cover how to create a simple Stimulus controller to reset the new post form.
```

Comment:
Thanks for that mention. It's not in the lesson yet, but I can include it, since it is a great example how to react to turbo lifecycle methods

Reply:
Unfortunately my Github settings seem to be a bit off, never got an email for being tagged in your completed thread last week.  Since your lesson is already approved and in a completed state, I can just change my reference to Stimulus to be more abstract in saying that the next lesson covers Stimulus and that's the tool you'd use to reset the element. I think that your Stimulus lesson flows from section-to-section in a way that teaches the content very well.

The post reset example is an extremely simple controller (simple being relative to other Stimulus controllers) and essentially just calling the Javascript `.reset()` function. The main difficulty is that this is sandwiched between the Foundations path and Javascript path and the learner has to actually check back to remember their DOM manipulation methods.  I think your lesson prepares them to use the framework extremely well and that this can be an oppritunity for self-learning homework for one who may want to implement it in a project later.